drivers/src/preconditioned_key.rs

```diff
@@ -0,0 +1,82 @@
+/*++
+
+Licensed under the Apache-2.0 license.
+
+File Name:
+
+    preconditioned_key.rs
+
+Abstract:
+
+    A FIPS-approved method for combining keys together leveraging a HMAC-based key-extraction process.
+
+--*/
+
+use crate::{
+    hmac_kdf, Aes, AesKey, AesOperation, Hmac, HmacData, HmacKey, HmacMode, HmacTag, KeyReadArgs,
+    KeyUsage, KeyWriteArgs, Trng,
+};
+
+use caliptra_error::{CaliptraError, CaliptraResult};
+
+pub fn preconditioned_key_extract(
+    input_key: HmacKey,
+    output_key: HmacTag,
+    label: &[u8],
+    salt: AesKey,
+    trng: &mut Trng,
+    hmac: &mut Hmac,
+    aes: &mut Aes,
+) -> CaliptraResult<()> {
+    let mut checksum = [0; 16];
+    aes.aes_256_ecb(salt, AesOperation::Encrypt, &[0; 16], &mut checksum)?;
+
+    match (salt, output_key) {
+        (AesKey::KV(salt_kv), HmacTag::Key(output_kv)) => {
+            hmac_kdf(
+                hmac,
+                input_key,
+                label,
+                Some(&checksum),
+                trng,
+                HmacTag::Key(KeyWriteArgs::new(
+                    output_kv.id,
+                    KeyUsage::default().set_hmac_key_en(),
+                )), // TODO(clundin): This re-uses the output key vault slot as a scratch space for the KDF. Verify this is okay and that it works.
+                HmacMode::Hmac512,
+            )?;
+
+            hmac.hmac(
+                HmacKey::Key(KeyReadArgs::new(output_kv.id)),
+                HmacData::Key(salt_kv),
```

Comment:
I believe these should be swapped. Ditto in the other switch case.

```suggestion
                HmacData::Key(salt_kv),
                HmacKey::Key(KeyReadArgs::new(output_kv.id)),
```

Reply:
In the other case I think this could be an issue:

`AesKey::Array` == 32 bytes
`HmacKey::Array` == 48 || 64 bytes

In that case, do you want to pad the AesKey so it can be used as an `HmacKey`?

Reply:
Ah. I see two options:

1. We do what you suggest and pad the AesKey when creating an HmacKey.
2. We change the type of `salt` to HmacKey, and then create an AesKey from that, truncating the bytes if needed. If a caller wants to pass a shorter salt, they can pad it themselves.

Option (2) would be needed if we ever want to pass a firmware-visible key larger than 256 bits as the salt. I don't think the LOCK spec currently has that. The only place that might qualify would be the initial MPK secret. The actual MPK secrets are all 512 bits, but the initial MPK secret, which comes from fw, would need to be 256 bits if `salt` remains an AesKey.

Reply:
After discussion we will add option 2 - I will modify the implementation to match this.